core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/registry/CachingCodecRegistry.java

```diff
@@ -273,15 +277,10 @@ private boolean collides(TypeCodec<?> newCodec, TypeCodec<?> oldCodec) {
       }
     }
 
-    if (value instanceof TupleValue) {
-      return uncheckedCast(codecFor(((TupleValue) value).getType(), TupleValue.class));
-    } else if (value instanceof UdtValue) {
-      return uncheckedCast(codecFor(((UdtValue) value).getType(), UdtValue.class));
-    }
-
-    GenericType<?> javaType = inspectType(value, null);
+    DataType cqlType = inferCqlTypeFromValue(value);
+    GenericType<?> javaType = inspectType(value, cqlType);
     LOG.trace("[{}] Continuing based on inferred type {}", logPrefix, javaType);
-    return uncheckedCast(getCachedCodec(null, javaType, true));
+    return uncheckedCast(getCachedCodec(cqlType, javaType, true));
```

Comment:
are we sure in this case that `true` should be passed as `isJavaCovariant` for all cases?

Reply:
`isJavaCovariant` should always be true if we are dealing with a real object. This is because your object is likely to be something like `ArrayList<Inet4Address>` but the registry would return a codec that handles `List<InetAddress>` instead.